### PR TITLE
Fixes #12414: Make reporting work when targeting several time the same item with several generic method

### PR DESCRIPTION
**Pull request by ncharles**
https://www.rudder-project.org/redmine/issues/12414

### CODE REVIEW

File: tree/20_cfe_basics/ncf_lib.cf [hunk 1]
```diff
@@ -879,14 +879,14 @@ body classes classes_generic_return_boolean_list_two(x, y, boolean_prefix, true_
...
-  !has_promiser_stack::
+#  !has_promiser_stack::
     promise_repaired => { "promise_repaired_$(x)", "$(x)_repaired", "$(x)_ok", "$(x)_reached" };
     repair_failed => { "repair_failed_$(x)", "$(x)_failed", "$(x)_not_ok", "$(x)_error", "$(x)_not_kept", "$(x)_not_repaired", "$(x)_reached" };
     repair_denied => { "repair_denied_$(x)", "$(x)_denied", "$(x)_not_ok", "$(x)_error", "$(x)_not_kept", "$(x)_not_repaired", "$(x)_reached" };
     repair_timeout => { "repair_timeout_$(x)", "$(x)_timeout", "$(x)_not_ok", "$(x)_error", "$(x)_not_kept", "$(x)_not_repaired", "$(x)_reached" };
     promise_kept => { "promise_kept_$(x)", "$(x)_kept", "$(x)_ok", "$(x)_not_repaired", "$(x)_reached" };
 
-  has_promiser_stack::
+#  has_promiser_stack::
```
Comment (ncharles) [inline tree/20_cfe_basics/ncf_lib.cf]: we should drop has_promiser_stack, and maybe define a class in the generic method that advertise that it use the class_prefix

**Comment by ncharles** [comment]
Commit modified

**Comment by amousset** [comment]
This is great!

File: tree/30_generic_methods/file_key_value_present_option.cf [hunk 2]
```diff
@@ -36,10 +36,11 @@ bundle agent file_key_value_present_option(file, key, value, separator, option)
 {
   vars:
       "old_class_prefix" string => canonify("file_key_value_present_${file}");
-      "promisers"          slist => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
-      "class_prefix"      string => canonify(join("_", "promisers"));
       "args"              slist => { "${file}", "${key}", "${value}","${separator}", "${option}" };
 
+      "report_param"      string => join("_", args);
+      "class_prefix"      string => canonify("${old_class_prefix}_${report_param}");
```
Comment (amousset) [inline tree/30_generic_methods/file_key_value_present_option.cf]: This may easily break if this string gets longer than 1024 chars

Comment (amousset) [reply tree/30_generic_methods/file_key_value_present_option.cf]: I think it fails clearly though (and could could check this in the technique editor)

Comment (amousset) [inline tree/30_generic_methods/file_key_value_present_option.cf]: This could fail if string is longer than 1024 chars.

**Comment by ncharles** [comment]
Commit modified

**Comment by ncharles** [comment]
Commit modified

File: tree/30_generic_methods/file_key_value_present_option.cf (hunk 2 above)
Comment (ncharles) [reply tree/30_generic_methods/file_key_value_present_option.cf]: this fails nicely: classes are not defined. However, we need in this case to catch the fact that classes are too long (see log_rudder.cf)

File: tree/20_cfe_basics/ncf_lib.cf [hunk 3]
```diff
@@ -879,14 +879,14 @@ body classes classes_generic_return_boolean_list_two(x, y, boolean_prefix, true_
 # Define x and y prefixed/suffixed with promise outcome
 body classes classes_generic_two(x,y)
 {
-  !has_promiser_stack::
+#  !has_promiser_stack::
     promise_repaired => { "promise_repaired_$(x)", "$(x)_repaired", "$(x)_ok", "$(x)_reached" };
```
Comment (peckpeck) [inline tree/20_cfe_basics/ncf_lib.cf]: You should drop those lines too since they duplicate what is done with has_promiser_stack below

Comment (ncharles) [reply tree/20_cfe_basics/ncf_lib.cf]: you are totally right !

**Comment by ncharles** [comment]
Commit modified

File: tree/20_cfe_basics/log_rudder.cf [hunk 4]
```diff
@@ -32,13 +32,34 @@ bundle agent log_rudder(message, old_class_prefix, origin_class_prefix, args)
```
Comment (amousset) [inline tree/20_cfe_basics/log_rudder.cf]: could you rename the parameters to make their meaning clearer?

**Comment by ncharles** [comment]
Generic methods code have been replaced with
for i  in `grep -L "disable_reporting" [a-z]*.cf`; do sed -i "/promisers/d" $i ;   done
for i  in `grep -L "disable_reporting" [a-z]*.cf`; do sed -i '/.*"args".*slist.*/a\      "report_param"      string => join("_", args);' $i ;   done
for i  in `grep -L "disable_reporting" [a-z]*.cf`; do sed -i '/.*join("_", args.*/a\      "class_prefix"      string => canonify("${old_class_prefix}_${report_param}");' $i ;   done

**Comment by ncharles** [comment]
Commit modified

**Comment by ncharles** [comment]
Commit modified

File: tree/20_cfe_basics/log_rudder.cf [hunk 5]
```diff
@@ -32,13 +32,29 @@ bundle agent log_rudder(message, old_class_prefix, origin_class_prefix, args)
...
 
-  methods:
+      "class_prefix_length"   int => string_length("${origin_class_prefix}");
+
+  classes:
+      # Use the full class prefix if:
+      # - it is defined
+      # - its length is smaller than 1000 character (classes are limited to 1023 characters, so if the prefix
+      #   is too long, then classes are not defined and it can't be used
+      # Note that if we don't use the full class_prefix, we fallback to the previous use case, where classes collision happen
```
[the remark this answers is not in the document]
Comment (ncharles) [reply tree/20_cfe_basics/log_rudder.cf]: done

File: tests/acceptance/30_generic_methods/log_rudder.cf [hunk 6]
```diff
@@ -13,7 +13,7 @@ bundle common acc_path
 body common control
 {
       inputs => { "${acc_path.root}/default.cf.sub", "${acc_path.root}/default_ncf.cf.sub", "@{ncf_inputs.default_files}" };
-      bundlesequence  => { configuration, default("${this.promise_filename}") };
+      bundlesequence  => { configuration, default("$(this.promise_filename)") };
```
Comment (amousset) [inline tests/acceptance/30_generic_methods/log_rudder.cf]: why changing this?

File: tests/acceptance/30_generic_methods/log_rudder.cf [hunk 7]
```diff
@@ -1,8 +1,8 @@
-#######################################################
+#########################################################################################
 #
-# Test checking if reporting is correcly done with rudder_logger
+# This will check reporting within log_rudder, by checking the defined classes
```
Comment (amousset) [inline tests/acceptance/30_generic_methods/log_rudder.cf]: Could you keep this test and make the class test a new one? It is good to keep a test of the reports output too.

Comment (ncharles) [reply tests/acceptance/30_generic_methods/log_rudder.cf]: Damn - i overwrote by mistake the test :/ i'll correct this ASAP

**Comment by ncharles** [comment]
Commit modified

**Pull request merged**